src/rn2xx3.cpp

```diff
@@ -97,88 +104,98 @@ bool rn2xx3::init()
   _nwkskey = "0";
-  _appskey = AppKey; //reuse the variable
   String receivedData;
 
   //clear serial buffer
   while(_serial.available())
     _serial.read();
 
-  _serial.println("sys get hweui");
-  String addr = _serial.readStringUntil('\n');
-  addr.trim();
-
   configureModuleType();
 
   switch (_moduleType) {
     case RN2903:
-      _serial.println("mac reset");
-      break;
+      sendRawCommand(F("mac reset"));
+    break;
     case RN2483:
-      _serial.println("mac reset 868");
-      break;
+      sendRawCommand(F("mac reset 868"));
+    break;
     default:
       // we shouldn't go forward with the init
       return false;
   }
-  receivedData = _serial.readStringUntil('\n');
 
-  _serial.println("mac set appeui "+_appeui);
-  receivedData = _serial.readStringUntil('\n');
+  // Key not empty, use it
+  if (DevEUI.length() == 16) {
+      _deveui = DevEUI;
+  } else {
+    // If not all keys are empty 
+    // In this case use on stored in RN2xx3 module
+    if (AppEUI!="" && AppKey!="")
+    {
+      String addr = sendRawCommand(F("sys get hweui"));
+      if(addr!="" && addr.length() == 16 )
+      {
+        _deveui = addr;
+      }
+    }
+  }
 
-  _serial.println("mac set appkey "+_appskey);
-  receivedData = _serial.readStringUntil('\n');
 
-  if(addr!="" && addr.length() == 16)
+  // Key not empty, use it, else use on stored in RN2xx3 module
+  if ( DevEUI.length() == 16 )
   {
-    _serial.println("mac set deveui "+addr);
+    sendRawCommand("mac set deveui "+_deveui);
   }
-  else
+
+  // Key not empty, use it, else use on stored in RN2xx3 module
+  if ( AppEUI.length() == 16 )
   {
-    _serial.println("mac set deveui "+_default_deveui);
+      _appeui = AppEUI;
+      sendRawCommand("mac set appeui "+_appeui);
+  }
+
+  // Key not empty, use it, else use on stored in RN2xx3 module
+  if ( AppKey.length() == 32 )
+  {
+    _appskey = AppKey; //reuse the variable
+    sendRawCommand("mac set appkey "+_appskey);
   }
-  receivedData = _serial.readStringUntil('\n');
 
   if (_moduleType == RN2903)
   {
-    _serial.println("mac set pwridx 5");
+    sendRawCommand("mac set pwridx 5");
   }
   else
   {
-    _serial.println("mac set pwridx 1");
+    sendRawCommand("mac set pwridx 1");
   }
-  receivedData = _serial.readStringUntil('\n');
 
-  _serial.println("mac set adr off");
-  receivedData = _serial.readStringUntil('\n');
+  sendRawCommand("mac set adr off");
 
   // Switch off automatic replies, because this library can not
   // handle more than one mac_rx per tx. See RN2483 datasheet,
   // 2.4.8.14, page 27 and the scenario on page 19.
-  _serial.println("mac set ar off");
-  _serial.readStringUntil('\n');
+  //sendRawCommand("mac set ar off");
 
   if (_moduleType == RN2483)
   {
-    _serial.println("mac set rx2 3 869525000");
-    receivedData = _serial.readStringUntil('\n');
+    sendRawCommand("mac set rx2 3 869525000");
   }
 
+  sendRawCommand("mac save");
+
   _serial.setTimeout(30000);
```

Comment:
We first need to adjust the timeout, then we can call `mac save`. Otherwise the serial.readStringUntil() will time out while waiting for the save to complete. Save is a slow operation.